src/engine/source/builder/src/builders/opmap/opBuilderHelperMap.cpp

```diff
@@ -2012,7 +2046,338 @@ MapOp opBuilderHelperIPVersionFromIPStr(const std::vector<OpArg>& opArgs,
+        }
+
+        const auto proto64 = protoOpt.value();
+        if (proto64 < 0 || proto64 > std::numeric_limits<uint8_t>::max())
+        {
+            throw std::runtime_error("network.iana_number out of range (expected 0..255)");
+        }
+
+        protoLiteral = static_cast<uint8_t>(proto64);
+    }
+
+    const auto name = buildCtx->context().opName;
+    const auto successTrace = fmt::format("{} -> Success", name);
+
+    const auto missingTrace = [name](const Reference& ref)
+    {
+        return fmt::format("{} -> Reference '{}' not found", name, ref.dotPath());
+    };
+    const auto stringTrace = [name](const Reference& ref)
+    {
+        return fmt::format("{} -> Reference '{}' must be a string", name, ref.dotPath());
+    };
+    const auto numberTrace = [name](const Reference& ref)
+    {
+        return fmt::format("{} -> Reference '{}' must be a number", name, ref.dotPath());
+    };
+    const auto invalidIPTrace = [name](const Reference& ref)
+    {
+        return fmt::format("{} -> Invalid IP format in '{}'", name, ref.dotPath());
+    };
+
+    const auto protoOutOfRangeTrace = fmt::format("{} -> network.iana_number out of range (expected 0..255)", name);
+    const auto sportOutOfRangeIcmpTrace = fmt::format("{} -> source.port out of range (expected 0..255)", name);
+    const auto dportOutOfRangeIcmpTrace = fmt::format("{} -> destination.port out of range (expected 0..255)", name);
+    const auto sportOutOfRangeTransportTrace = fmt::format("{} -> source.port out of range (expected 0..65535)", name);
+    const auto dportOutOfRangeTransportTrace =
+        fmt::format("{} -> destination.port out of range (expected 0..65535)", name);
+    const auto invalidIpTrace = [name](const Reference& ref)
+    {
+        return fmt::format("{} -> Invalid IP format in '{}'", name, ref.dotPath());
+    };
+
+    const auto computeFailureTrace = fmt::format("{} -> failed to compute Community ID", name);
+    const auto buildBufferFailureTrace = fmt::format("{} -> failed to build Community ID buffer", name);
+    const auto sha1FailureTrace = fmt::format("{} -> SHA1 digest computation failed", name);
+    const auto base64FailureTrace = fmt::format("{} -> Base64 encoding failed", name);
+
+    return [=, runState = buildCtx->runState()](base::ConstEvent event) -> MapResult
+    {
+        std::string saddr;
+        std::string daddr;
+
+        auto readIp = [&](const Reference& ref,
+                          const std::string& missingMsg,
+                          const std::string& typeMsg,
+                          const std::string& invalidMsg,
+                          std::string& out) -> std::optional<std::string>
+        {
+            const auto valueOpt = event->getString(ref.jsonPath());
+            if (!valueOpt)
+            {
+                if (!event->exists(ref.jsonPath()))
+                {
+                    return missingMsg;
+                }
+                return typeMsg;
+            }
```

Comment:
Use a single message to indicate whether the value does not exist or is not a string, to avoid using the exists() function.